libnetwork/internal/resolvconf/resolvconf.go

```diff
@@ -508,3 +492,19 @@ func defaultNSAddrs(ipv6 bool) []netip.Addr {
 	}
 	return addrs
 }
+
+func (rc *ResolvConf) removeInvalidNDots() {
+	n := 0
+	for _, opt := range rc.options {
+		k, v, _ := strings.Cut(opt, ":")
+		if k == "ndots" {
+			ndots, err := strconv.Atoi(v)
+			if err != nil || ndots < 0 {
+				continue
+			}
+		}
+		rc.options[n] = opt
+		n++
+	}
+	rc.options = rc.options[:n]
+}
```

Comment:
This could be written as a pure(-ish) function. I like pure functions as side effect-free code is much easier to reason about. Also, when reusing a slice's backing array, the elements between len and cap are considered live references. Strings are a reference type, so the elements have to be zeroed out to allow the strings to be GC'ed.
```suggestion
// removeInvalidNDots filters ill-formed "ndots" settings from options.
// The backing array of the options slice is reused.
func removeInvalidNDots(options []string) []string {
	n := 0
	for _, opt := range options {
		k, v, _ := strings.Cut(opt, ":")
		if k == "ndots" {
			ndots, err := strconv.Atoi(v)
			if err != nil || ndots < 0 {
				continue
			}
		}
		options[n] = opt
		n++
	}
	clear(options[n:]) // Zero out the obsolete elements, for GC.
	return options[:n]
}
```
(The `clear` builtin is new in Go 1.21, and because it's a builtin we can use it without getting burned by the Go 1.21 `//go:build` generics bugs.)

Reply:
Thanks Cory - done.

Reply:
The "SliceTricks" wiki / write-up is also always a good read; https://go.dev/wiki/SliceTricks#filtering-without-allocating

> Strings are a reference type, so the elements have to be zeroed out to allow the strings to be GC'ed.

@corhere would that be something worth adding to the example in SliceTricks? https://go.dev/wiki/SliceTricks#filter-in-place

Reply:
> would that be something worth adding to the example in SliceTricks?

Probably not, because [`slices.DeleteFunc`](https://pkg.go.dev/slices#DeleteFunc) exists and anyone who isn't us would be better off using that instead.